src/runtime/virtcontainers/qemu_amd64.go

```diff
@@ -273,3 +278,129 @@ func (q *qemuAmd64) appendProtectionDevice(devices []govmmQemu.Device, firmware
+		"kernel-params-non-debug": q.kernelParamsNonDebug,
+		"kernel-params":           q.kernelParams})
+	switch q.protection {
+	case sevProtection:
+		logger.Info("SEV attestation: Pulling launch argument...")
+		logger.Info("SEV attestation: Server %s", proxy)
+		logger.Info("SEV attestation: Path %s", path)
+
+		// start VM in stalled state
+		config.Knobs.Stopped = true
+
+		// Pull the launch blob and godh from GOP
+		// Use an external script until GOP client logic is built into the kata-runtime
+		cmd := exec.Command(sevGuestOwnerProxyClient, "GetBundle", proxy, path)
+		logger.Info("SEV attestation: GetBundle Command: ", cmd.String())
+		out, err := cmd.CombinedOutput()
+		cmd.Wait()
+		if err != nil {
+			logger.Info("SEV attestation: GetBundle FAILED: %s", err)
+			return config, err
+		}
+		// TODO: error check
+		out_string := strings.TrimSuffix(string(out), "\n")
+		logger.Info("SEV attestation: Received GOP results: %s", out_string)
+		gop_result := strings.Split(string(out_string), ",")
+		fmt.Println(gop_result)
+		logger.Info("SEV attestation: godh %s", gop_result[0])
+		logger.Info("SEV attestation: launch measure %s", gop_result[1])
+		logger.Info("SEV attestation: connection id %s", gop_result[2])
+
+		// Place launch args into qemuConfig.Devices struct
+		for i := range config.Devices {
+			if reflect.TypeOf(config.Devices[i]).String() == "qemu.Object" {
+				if config.Devices[i].(govmmQemu.Object).Type == govmmQemu.SEVGuest {
+					logger.Info("SEV attestation: UPDATING DEVICE")
+					dev := config.Devices[i].(govmmQemu.Object)
+					dev.CertFilePath = gop_result[0]
+					dev.SessionFilePath = gop_result[1]
+					dev.DeviceID = gop_result[2]
+					dev.SevPolicy = 0
+					config.Devices[i] = dev
+					break
+				}
+			} else {
+				logger.Info("SEV attestation: ELSE %s", reflect.TypeOf(config.Devices[i]).String())
+			}
+		}
+		return config, nil
+	default:
+		return config, nil
+	}
+}
+
+// wait for prelaunch attestation to complete
+func (q *qemuArchBase) prelaunchAttestation(ctx context.Context, qmp *govmmQemu.QMP, config govmmQemu.Config, path string, proxy string, keyset string) error {
+	logger := virtLog.WithFields(logrus.Fields{
+		"subsystem":               "qemuAmd64",
+		"machine":                 q.qemuMachine,
+		"kernel-params-debug":     q.kernelParamsDebug,
+		"kernel-params-non-debug": q.kernelParamsNonDebug,
+		"kernel-params":           q.kernelParams})
+	switch q.protection {
+	case sevProtection:
+		// This will block and wait for the Guest Owner Proxy to validate the
+		// prelaunch attestation measurement.
+		logger.Info("SEV attestation: Processing prelaunch attestation")
+		logger.Info("SEV attestation: Server %s", proxy)
+		logger.Info("SEV attestation: Path %s", path)
+		logger.Info("SEV attestation: Keyset %s", keyset)
+		connection_id := ""
+		var sev_policy uint32 = 0
+		for i := range config.Devices {
+			if reflect.TypeOf(config.Devices[i]).String() == "qemu.Object" {
```

Comment:
- I thought reflect had a relatively significant performance impact. Maybe that's no longer the case but do we really *have* to use this package?
- What if `qemu.Object` changes name?

Reply:
Agreed - this is very ugly. But, I'm not really sure another way to go about it. What is happening here is that we  traverse a list of generic-typed Device objects to get a reference to a previously-created `sev-guest` QEMU object. We assume there's only one `qemu.Object` on the list, but that's not necessarily true. 

> I thought reflect had a relatively significant performance impact.
Not sure. We traverse the device list and call reflect twice during the VM setup and launch. The attestation process is already pretty expensive and the device list is very short, O(1), so I haven't noticed a performance hit. 

> What if qemu.Object changes name? 
And what if there's more than one `qemu.Object`? This piece of code is full of dangerous assumptions. 

Any thoughts @Jakob-Naucke ?

Reply:
> Any thoughts @Jakob-Naucke ?

I wouldn't know of another way unless we introduce something annotation-like to govmm (which would be too many LOC to justify it IMO).

> what if there's more than one `qemu.Object`?

Aren't you still checking all elements individually?

Also, I've been trying to play with this, and the signature of `ExecuteSEVInjectLaunchSecret` doesn't match confidential-containers-demo/govmm/sev-attestation. Does this need an update?

Reply:
>  Does this need an update?

Yup, nice catch! All synced